### PR TITLE
Track privacy head

### PR DESCRIPTION
**Pull request by iikirilov**
## PR description

Creates and registers `BlockAddedObserver` which will track the latest block that has a PMT for every privacy group at every block.

I.e. when a new block is added - a map with `privacyGroupId -> lastBlockWithPMT` is created. The map is only updated for those privacyGroups that have had a PMT in the block that had just been added.

### CODE REVIEW

**Comment by lucassaldanha** [comment]
I think the problem of using a BlockAddedObserver is that the observer callbacks are executed asynchronously. So we can't guarantee that we have the map ready to be used while another block is being processed. We need something that will guarantee that after we process the block, we update the mapping, and after that we are ready to process any other blocks.

One idea that we had was to create a PrivateBlockProcessor (implementing BlockProcessor) and wrap the MainnetBlockProcessor (kinda of a proxy). This way, we can delegate the block processing to the MainnetBlockProcessor, and after the MainnetBlockProcessor#processBlock finishes, we can update the mapping.

The idea is that it would give us entry points before and after processing the block to add any logic that we need.

What do you think?

**Comment by iikirilov** [comment]
Yep, I understand. I forgot it was asynchronous. It is just a matter of refactoring what is currently here :)

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/precompiles/privacy/PrivacyPrecompiledContract.java [hunk 1]
```diff
@@ -109,10 +115,41 @@ public BytesValue compute(final BytesValue input, final MessageFrame messageFram
 
     final BytesValue privacyGroupId = BytesValues.fromBase64(receiveResponse.getPrivacyGroupId());
 
-    // get the last world state root hash - or create a new one
-    final Hash lastRootHash =
-        privateStateRootResolver.resolveLastStateRoot(
-            messageFrame.getBlockchain(), messageFrame.getBlockHeader(), privacyGroupId);
+    final Map<Bytes32, Hash> privacyGroupToLatestBlockWithTransactionMap =
```
Comment (pinges) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/precompiles/privacy/PrivacyPrecompiledContract.java]: I'd prefer if we were to get the map only once per block, update the map while we are processing the private transactions, and then persist it when we finished the block.

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/storage/PrivateGroupIdToLatestBlockWithTransactionMap.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+import org.hyperledger.besu.ethereum.rlp.RLPInput;
+import org.hyperledger.besu.ethereum.rlp.RLPOutput;
+import org.hyperledger.besu.util.bytes.Bytes32;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+public class PrivateGroupIdToLatestBlockWithTransactionMap {
```
Comment (pinges) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/storage/PrivateGroupIdToLatestBlockWithTransactionMap.java]: Someone might have a better shorter name for the class ...

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/precompiles/privacy/PrivacyPrecompiledContract.java (hunk 1 above)
Comment (iikirilov) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/precompiles/privacy/PrivacyPrecompiledContract.java]: that works

Comment (iikirilov) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/precompiles/privacy/PrivacyPrecompiledContract.java]: we can pass it in nicely to the MessageFrame - I have gone for a middle ground where we copy the privacy group map before processing the block and only update it if and when we process a PMT in the block.

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/storage/PrivateGroupIdToLatestBlockWithTransactionMap.java (hunk 2 above)
Comment (iikirilov) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/storage/PrivateGroupIdToLatestBlockWithTransactionMap.java]: sorted

**Comment by lucassaldanha** [comment]
I believe we still need to update `ChainHeadPrivateNonceProvider` to stop using the state root resolver. It is causing the thread block issue.

**Pull request merged**